tests/unit/s2n_x509_validator_test_offset.c

```diff
@@ -0,0 +1,156 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+static int mock_time(void *data, uint64_t *timestamp)
+{
+    *timestamp = *(uint64_t *) data;
+    return 0;
+}
+
+static int read_file(struct s2n_stuffer *file_output, const char *path, uint32_t max_len)
+{
+    FILE *fd = fopen(path, "rb");
+    s2n_stuffer_alloc(file_output, max_len);
+
+    if (fd) {
+        char data[1024];
+        size_t r = 0;
+        while ((r = fread(data, 1, sizeof(data), fd)) > 0) {
+            s2n_stuffer_write_bytes(file_output, (const uint8_t *) data, (const uint32_t) r);
+        }
+        fclose(fd);
+        return s2n_stuffer_data_available(file_output) > 0;
+    }
+
+    return -1;
+}
+
+struct host_verify_data {
+    const char *name;
+    uint8_t found_name;
+    uint8_t callback_invoked;
+};
+
+static uint8_t verify_host_accept_everything(const char *host_name, size_t host_name_len, void *data)
+{
+    return 1;
+}
+
+int test_with_system_time(uint64_t system_time_nanos, int expected)
+{
+    DEFER_CLEANUP(struct s2n_x509_trust_store trust_store = { 0 }, s2n_x509_trust_store_wipe);
+    s2n_x509_trust_store_init_empty(&trust_store);
+    EXPECT_SUCCESS(s2n_x509_trust_store_from_ca_file(&trust_store, S2N_OCSP_CA_CERT, NULL));
+
+    DEFER_CLEANUP(struct s2n_x509_validator validator = { 0 }, s2n_x509_validator_wipe);
+    EXPECT_SUCCESS(s2n_x509_validator_init(&validator, &trust_store, 1));
+
+    struct s2n_connection *connection = s2n_connection_new(S2N_CLIENT);
+    EXPECT_NOT_NULL(connection);
+
+    struct host_verify_data verify_data = { .callback_invoked = 0, .found_name = 0, .name = NULL };
+    EXPECT_SUCCESS(s2n_connection_set_verify_host_callback(connection, verify_host_accept_everything, &verify_data));
+
+    DEFER_CLEANUP(struct s2n_stuffer cert_chain_stuffer = { 0 }, s2n_stuffer_free);
+    EXPECT_OK(s2n_test_cert_chain_data_from_pem(connection, S2N_OCSP_SERVER_CERT_EARLY_EXPIRE, &cert_chain_stuffer));
+    uint32_t chain_len = s2n_stuffer_data_available(&cert_chain_stuffer);
+    uint8_t *chain_data = s2n_stuffer_raw_read(&cert_chain_stuffer, chain_len);
+    EXPECT_NOT_NULL(chain_data);
+
+    DEFER_CLEANUP(struct s2n_pkey public_key_out = { 0 }, s2n_pkey_free);
+    EXPECT_SUCCESS(s2n_pkey_zero_init(&public_key_out));
+    s2n_pkey_type pkey_type = S2N_PKEY_TYPE_UNKNOWN;
+    EXPECT_OK(s2n_x509_validator_validate_cert_chain(&validator, connection, chain_data, chain_len, &pkey_type, &public_key_out));
```

Comment:
Is there any reason for calling validate_cert_chain in this test?

Reply:
Yes, otherwise I was getting errors about some "unexpected validator state" or something like that.